tests/migration_test.go

```diff
@@ -930,25 +931,67 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 			})
 		})
 		Context("with an Fedora shared NFS PVC, cloud init and service account", func() {
-			var pvName string
 			var vmi *v1.VirtualMachineInstance
-			BeforeEach(func() {
+			var pvName string
+			var nfsInitialized bool
+
+			initNFS := func(ipFamily k8sv1.IPFamily) string {
 				tests.SkipNFSTestIfRunnigOnKindInfra()
-				pvName = "test-nfs" + rand.String(48)
+				pvName := "test-nfs" + rand.String(48)
 				// Prepare a NFS backed PV
 				By("Starting an NFS POD")
 				os := string(cd.ContainerDiskFedora)
-				nfsIP := tests.CreateNFSTargetPOD(os)
+				nfsPod := tests.CreateNFSTargetPOD(os)
+
 				// create a new PV and PVC (PVs can't be reused)
 				By("create a new NFS PV and PVC")
+				nfsIP := libnet.GetPodIpByFamily(nfsPod, ipFamily)
+				ExpectWithOffset(1, nfsIP).NotTo(BeEmpty())
 				tests.CreateNFSPvAndPvc(pvName, "5Gi", nfsIP, os)
+				nfsInitialized = true
+				return pvName
+			}
+
+			AfterEach(func() {
+				if vmi != nil {
+					By("Deleting the VMI")
+					Expect(virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})).To(Succeed())
+
+					By("Waiting for VMI to disappear")
+					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+
+					vmi = nil
+				}
 			})
 
 			AfterEach(func() {
-				// PVs can't be reused
-				tests.DeletePvAndPvc(pvName)
+				if nfsInitialized {
```

Comment:
Is the function that initializes NFS atomic? And if not, does that function clean up if it fails in the middle of its execution? If neither of those apply, we may leave the cluster in a broken state after failed initialization of resources.

The same applies for all the other AfterEach conditions.

Reply:
You are right. If the `initNFS` fail in the middle then the `AfterEach` methods won't clean the resource that the `initNFS` already created.

To avoid over complication of the tests code. And reduce the risk of leaving garbage after the test run.
Can we agree on the following approach (I call it "unconditional `AfterEach`es")? (This PR is just an example but this is a suggestion for a general way to go. Especially for our new ipv6 tests)
1. **Clean** resources of a **skipped** test in an **unconditional** way.
The `AfterEach` methods of a skipped test are invoked but if they have errors, they are not shown and doesn't affect the result of the test.
* Resources created by the test - will be removed.
* Resources not created by the test - will get a silent error.

2. **Clean** resources of a **failed** test in an **unconditional** way.
* Resources created by the test - will be removed.
* Resources not created by the test - will get a an error, but the test failed anyway so the cleanup error can be considered as silent.

Dealing with all those `AfterEach` and cleanup dilemmas made me to think that maybe we are over-thinking this issue and complicating the tests.

We have `utils.BeforeTestCleanup` and `utils.AfterTestSuitCleanup` which clean the test namespace after each test (yes, I know, it is before the next test run but if has the same affect).
Unless we have to wait for the resource to actually be cleaned before the next test run, why not rely on this general mechanism?

@EdDev @phoracek @maiqueb

Reply:
> We have utils.BeforeTestCleanup and utils.AfterTestSuitCleanup which clean the test namespace after each test (yes, I know, it is before the next test run but if has the same affect).

Unfortunately, it is does not have the same effect:
- Cleaning before assumes the next test is performing such a cleanup, which may not be correct. A simple example is a single focus test.
- Cleaning the whole namespace instead of individual resources cleans **everything**, both masking bugs and limiting re-usage of resources between tests (although some argue this is not needed, something I disagree).
- When we introduce a new object that needs cleanup, it must be explicitly added to theses cleanup functions otherwise they will not get cleaned. It happened already: The Jobs objects.

Regarding the skipped tests: The skips could be executed on a dedicated `BeforeEach` on the top most container that requires it. Then, no resources need to be created at all.

Setup/Teardown:
-  There should be one setup per resource. A flag that attempts to aggregate multiple setups will not work that well and will most likely complicate things.
- Conditional cleanup can work on container variables check, however, there is a need to check that the values are not leaking between tests. Initializing them on setup may do the job.
   - An unconditional cleanup may work as well, however, it will mess up the logs and I prefer to avoid it in case we find a cleaner solution.
- We could build a small generic infrastructure to deal with this, but this needs a broader discussion.
   - Something like this: A setup will return a clean func and an error. The error is asserted and if that passes we register the clean func into a cleanup slice. One teardown cleaner will pass the slice and execute it one by one, collecting the errors and reporting. We will not run cleanup in cases the setup fail or did not occur and it should be simple to init the slice post processing it.